### ISSUE TITLE
Equals method now not working

### ISSUE THREAD

**Issue by ArkAzrael** [nus-cs2030/2324-s2#418]
# Summary

Equals method for lab 5 worked before regarding, now it fails test case 7

## Description:

I submitted lab 5 but now my equals method doesn't seem to work properly for some reason after it was rereleased for regrading. It is returning false on test cases where it should be true. I have absolutely no idea what changed about it to cause this issue now.

[
![equals](https://github.com/nus-cs2030/2324-s2/assets/157274977/ddbefb8c-33bb-4f12-94e9-f7130793fc1b)
](url)

**Comment by niclow236** [comment]
Maybe you could try using implicit typecasts rather than explicit typecasts.

**Comment by joshualek09** [comment]
Hi ArkAzrael, 
i think this statement is problematic:
else if (obj instanceof Log) { Log other = (Log<?>) obj;

1. Log other should be changed to Log<?> other 
- you are performing the casting correctly with (Log<?>), but the type of other cannot be determined at runtime. So at runtime, Java Virtual Machine doesnt know that other is supposed to be a Log<?>. It only knows that it is a Log object. 
Put simply, if you dont use <?>, 
means there is no type safety, ie you dk the actual type argument,
thus, cannot handle the object properly due to issues at runtime

**Comment by ArkAzrael** [comment]
Hi I did specify the Log<?> other but it wasn't showing up with text. No idea why this doesn't work now when it did prior. Would it be the value part throwing the error?

**Comment by ArkAzrael** [comment]
I am unable to get the right output for the monad left identity and associativity

**Comment by ArkAzrael** [comment]
I managed to fix it by checking for empty strings before

**Comment by joshualek09** [comment]
im happy to hear that. 

im not sure if you did this following edit that resolved your issue, but your flatMap parameter could be further improved on. (but if you did it then thats great, feel free to ignore my comment hahah)

you can make it as:
 public <R> Log<R> flatMap(Function<? super T, ? extends Log<? extends R>> mapper)

 the difference lies with: ? extends Log<? extends R>
 it seems like your argument does not consider if the result of the lambda expression is a subtype of Log.

you may refer to lecture 7 video at the 42min mark, prof chia actually explains this situation. he also shows what should be the actual code: flatMap(Function<? super T, ? extends Log<? extends R>> mapper)

i recorded down the essence of what he said, so i hope it helps:
// this flatMap has "? extends Maybe<? extends R>", compared to just simply "Maybe<? extends R>"
// reason for this is because:
// 1) Maybe<? extends R> means that it is a Maybe object with type of R or any subtype of R
// 2) if you try to perform: Maybe.of(1).map(x -> Maybe.of(x+1)), then the result of the lambda has to be a Maybe. it cannot be a subtype of Maybe
// so what if we want a Maybe, or a subtype of Maybe? we will have to resolve the invariance, so have to use the ? extends wildcard. Hence, you do ? extends Maybe<? extends R>


if my explanation is not clear enough, perhaps this explanation by niclow236 is clearer:
The issue here is with the return type of the Function parameter and the expectation of it to return a Log<R>. Function<? super T, Log<R>> won't work but it should be Function<? super T, ? extends Log<? extends R>> instead.

? extends Log<? extends R> allows the function to return a Log or any subtype of Log that extends R, which is important for flexibility. The flatMap method wants to accommodate mappers that might return Log<R> or any subtype of it. If you constrain it to Log<R>, you might lose out on the ability to use mappers that return subtypes of Log<R>.

By allowing the mapper function to return Log<? extends R>, you're providing flexibility. For example, if R is String, the mapper function can return Log<String>, Log<Object>, or any subtype of Log<String>.

However, if you restrict it to Function<? super T, Log<R>>, you are saying that the mapper must always return exactly Log<R>.
